curvefs/src/mds/topology/topology.cpp

```diff
@@ -1129,129 +1133,118 @@ TopoStatusCode TopologyImpl::UpdatePartitionTxIds(
+            continue;
         }
-    }
 
-    if (metaServerList.empty()) {
-        LOG(ERROR) << "Choose metaserver in zone failed,"
-                   << " the metaserver is empty."
-                   << "zoneId = " << zoneId;
-        return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
-    }
+        Server server;
+        ServerIdType serverId = it->GetServerId();
+        if (GetServer(serverId, &server)) {
+            PoolIdType poolId = server.GetPoolId();
+            ZoneIdType zoneId = server.GetZoneId();
+            auto iter = candidateMap.find(poolId);
+            if (iter == candidateMap.end()) {
+                std::map<ZoneIdType, MetaServer> tmpMap;
+                tmpMap.emplace(zoneId, *it);
+                candidateMap.emplace(poolId, tmpMap);
+            } else {
+                bool isDup = false;
+                for (const auto &it : iter->second) {
+                    if (zoneId == it.first) {
+                        isDup = true;
+                        break;
+                    }
+                }
+                if (!isDup) {
+                    iter->second.emplace(zoneId, *it);
+                }
+            }
+        } else {
+            LOG(ERROR) << "get server failed when choose metaservers,"
+                       << " the serverId = " << serverId;
+            return TopoStatusCode::TOPO_SERVER_NOT_FOUND;
+        }
 
-    while (!metaServerList.empty()) {
-        auto iter = metaServerList.begin();
-        int randomValue = GetOneRandomNumber(0, metaServerList.size() - 1);
-        std::advance(iter, randomValue);
-        // check metaserver online status
-        MetaServer metaserver;
-        if (GetMetaServer(*iter, &metaserver)) {
-            // TODO(wanghai01): consider the space used by per copyset more
-            // details later  // NOLINT
-            uint64_t leftSpace =
-                metaserver.GetMetaServerSpace().GetDiskCapacity() -
-                metaserver.GetMetaServerSpace().GetDiskUsed();
-            if (ONLINE == metaserver.GetOnlineState() && leftSpace > 0) {
-                *metaServerId = *iter;
+        for (const auto &it : candidateMap) {
+            uint16_t replicaNum = 0;
+            auto iter = replicaMap.find(it.first);
+            if (iter == replicaMap.end()) {
+                Pool pool;
+                if (GetPool(it.first, &pool)) {
+                    replicaNum = pool.GetReplicaNum();
+                    replicaMap.emplace(it.first, replicaNum);
+                } else {
+                    LOG(ERROR) << "get pool failed when choose metaservers,"
+                               << " the poolId = " << it.first;
+                    return TopoStatusCode::TOPO_POOL_NOT_FOUND;
+                }
+            } else {
+                replicaNum = iter->second;
+            }
+            // find the available metaserver group
+            if (it.second.size() == replicaNum) {
```

Comment:
what if if (it.second.size() >= replicaNum)

Reply:
> what if if (it.second.size() >= replicaNum)

I think it makes no difference.